src/ucs/debug/log.c

```diff
@@ -94,34 +115,127 @@ void ucs_log_flush()
 size_t ucs_log_get_buffer_size()
 {
     return ucs_config_memunits_get(ucs_global_opts.log_buffer_size,
-                                   256, 2048);
+                                   256, UCS_ALLOCA_MAX_SIZE);
+}
+
+static void ucs_log_get_file_name(char *log_file_name, size_t max, unsigned idx)
+{
+    ucs_assert(idx <= ucs_global_opts.log_file_backup_count);
+
+    if (idx == 0) {
+        ucs_strncpy_zero(log_file_name, ucs_log_file_names[0], max);
+        return;
+    }
+
+    ucs_snprintf_zero(log_file_name, max, "%s.%u",
+                      ucs_log_file_names[idx], idx);
+}
+
+static void ucs_log_handle_file_max_size(int log_entry_len)
+{
+    const char *next_token;
+    char old_log_file_name[UCS_LOG_MAX_FILE_NAME];
+    char new_log_file_name[UCS_LOG_MAX_FILE_NAME];
+    unsigned idx, start_idx;
+    int ret;
+
+    /* check if it is necessary to find a new storage for logs */
+    if ((log_entry_len + ftell(ucs_log_file)) >=
+        ucs_global_opts.log_file_max_size) {
+        fclose(ucs_log_file);
+
+        if (ucs_global_opts.log_file_backup_count != 0) {
+            if (ucs_log_file_last_idx == ucs_global_opts.log_file_backup_count) {
+                /* remove the last file and log rotation from the
+                 * `log_file_backup_count - 1` file */
+                ucs_log_get_file_name(old_log_file_name,
+                                      UCS_LOG_MAX_FILE_NAME,
+                                      ucs_log_file_last_idx);
+                unlink(old_log_file_name);
+                start_idx = ucs_log_file_last_idx - 1;
+            } else {
+                start_idx = ucs_log_file_last_idx;
+                ucs_log_file_last_idx++;
+            }
+
+            ucs_assert(ucs_log_file_last_idx <=
+                       ucs_global_opts.log_file_backup_count);
+            ucs_assert(start_idx <= (ucs_global_opts.log_file_backup_count - 1));
+
+            for (idx = start_idx; idx != UINT_MAX; --idx) {
+                ucs_assert(idx <= start_idx);
+
+                ucs_log_get_file_name(old_log_file_name,
+                                      UCS_LOG_MAX_FILE_NAME, idx);
+                ucs_snprintf_zero(new_log_file_name, UCS_LOG_MAX_FILE_NAME,
+                                  "%s.%u", ucs_log_file_names[idx], idx + 1);
```

Comment:
why do we need to pre-allocate this array of names?
can't we just create it on the fly when needed?

Reply:
yes, we can
but if user specifies `%t `to the name of the log file, it wouldn't work, since the name has to be unchanged, we just need to add 1/2/.../N at the end of the old file name

Reply:
so we just need to keep the result of "%t", e.g the "base name", not all the names 0..n-1

Reply:
do you mean to change this array to keep only `%t` result?

Reply:
i think it does not have to be an array, just one string for base log file name, including %t result if needed

Reply:
after the first rotation, we will need to open the new file and we will generate the new name for it

Reply:
but you said the name has to be unchanged.. ?

Reply:
> but you said the name has to be unchanged.. ?

 I meant that unchanged when we backuping it, e.g.:
1. initially we have
```
log_file_10-30-20
```
2. after first rotation (in 2 minutes and 3 seconds):
```
log_file_10-32-23
log_file_10-30-20.1
```
3. after second rotation (in 3 minutes and 10 seconds):
```
log_file_10-35-33
log_file_10-32-23.1
log_file_10-30-20.2
```

Reply:
or we should keep the first name, i.e.:
```
log_file_10-30-20
log_file_10-30-20.1
log_file_10-30-20.2
```
could you please advise me which option works better for you?

Reply:
i think we should just keep first name, e.g `log_file_10-30-20`

Reply:
> i think we should just keep first name, e.g `log_file_10-30-20`

@yosefe makes sense to me too, will re-implement
thank you for the good catch!

Reply:
done